engine/runtime-compiler/src/main/java/org/enso/compiler/pass/lint/unusedimports/UnusedImports.java

```diff
@@ -0,0 +1,217 @@
+package org.enso.compiler.pass.lint.unusedimports;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.enso.compiler.context.InlineContext;
+import org.enso.compiler.context.ModuleContext;
+import org.enso.compiler.core.CompilerError;
+import org.enso.compiler.core.IR;
+import org.enso.compiler.core.ir.Expression;
+import org.enso.compiler.core.ir.Module;
+import org.enso.compiler.core.ir.Name.Literal;
+import org.enso.compiler.core.ir.Warning;
+import org.enso.compiler.core.ir.Warning.UnusedImport;
+import org.enso.compiler.core.ir.Warning.UnusedSymbolsFromImport;
+import org.enso.compiler.core.ir.module.scope.Import;
+import org.enso.compiler.core.ir.module.scope.imports.Polyglot;
+import org.enso.compiler.data.BindingsMap;
+import org.enso.compiler.data.BindingsMap.ResolvedName;
+import org.enso.compiler.pass.IRPass;
+import org.enso.compiler.pass.IRProcessingPass;
+import org.enso.compiler.pass.analyse.AmbiguousImportsAnalysis;
+import org.enso.compiler.pass.analyse.BindingAnalysis$;
+import org.enso.compiler.pass.analyse.ImportSymbolAnalysis;
+import org.enso.compiler.pass.resolve.GenericAnnotations$;
+import org.enso.compiler.pass.resolve.GlobalNames$;
+import org.enso.compiler.pass.resolve.MethodDefinitions;
+import org.enso.compiler.pass.resolve.Patterns$;
+import org.enso.compiler.pass.resolve.TypeNames$;
+import org.enso.compiler.pass.resolve.TypeSignatures$;
+import org.enso.pkg.QualifiedName;
+import org.enso.scala.wrapper.ScalaConversions;
+import scala.collection.immutable.Seq;
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * Attaches warnings to those {@link Import import IRs} that export symbols that are not used in any
+ * expression in the current module.
+ *
+ * <p>Works with then {@link BindingsMap.Resolution} metadata from various passes.
+ *
+ * <p>Traverses the entire {@link Module module IR}, looks for all the {@link BindingsMap.Resolution
+ * resolved} symbols, and assigns these resolutions to their corresponding {@link Import import IR}.
+ *
+ * <p>Intentionally not implemented as {@link org.enso.compiler.pass.MiniIRPass mini pass}.
+ *
+ * <p>This is an expensive pass and should be run only during linting.
```

Comment:
- having an expensive pass is unfortunate, but acceptable during linting
- having an expensive pass implemented via _old school_ traversal method is acceptable too